File: JS-Script/hosthunter.js
const dns = require('dns');
const fs = require('fs');
const chalk = require('chalk');
const path = require('path');

// Utility function to generate a random color for the tool name
const getRandomColor = () => {
    const colors = [
        chalk.red, chalk.green, chalk.yellow, chalk.cyan, chalk.magenta, chalk.blue
    ];
    return colors[Math.floor(Math.random() * colors.length)];
};

// Function to get the terminal width (cross-platform)
const getTerminalWidth = () => {
    return process.stdout.columns || 80; // Default to 80 if unable to get terminal width
};

// Validation function to check if a single IP address is valid
const isValidIp = (ip) => {
    const ipRegex = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
    return ipRegex.test(ip);
};

// Validation function to check if a CIDR IP range is valid
const isValidIpRange = (range) => {
    const rangeRegex = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\/([1-9]|[1-2][0-9]|3[0-2]))$/;
    return rangeRegex.test(range);
};

// Function to validate a list of IP addresses or ranges
const validateIps = (ips) => {
    return ips.every(ipLine => {
        // Split comma-separated IPs/ranges on each line
        const ipParts = ipLine.split(',').map(part => part.trim());
        
        // Validate each part individually
        return ipParts.every(ip => {
            const isValid = isValidIp(ip) || isValidIpRange(ip);
            if (!isValid) {
                console.error(chalk.red(`Invalid IP address or range format: "${ip}"`));
            }
            return isValid;
        });
    });
};

// Variable Initialization
const args = process.argv.slice(2);
let ipAddresses = [];
let outputFileName = null;
let results = [['IP Address', 'Status', 'Hostname']]; // Initial headers for results
let completedLookups = 0; // Track completed lookups

// Stylized output for HostHunter tool name with random color
const randomColor = getRandomColor();
const toolName = `
${randomColor.bold('HHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH')}
${randomColor.bold('H')}${randomColor('              HostHunter              ')}${randomColor.bold('H')}
${randomColor.bold('H')}${randomColor('      A Reverse DNS Lookup Tool       ')}${randomColor.bold('H')}
${randomColor.bold('HHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH')}
`;

// Display the tool name at the start of the script
console.log(toolName);

// Section 2: Argument Parsing and File Validation

// Argument Parsing and Error Handling
for (let i = 0; i < args.length; i++) {
    if (args[i] === '-i') {
        const input = args[i + 1];
        ipAddresses = input.split(',').map(ip => ip.trim());
        i++;
    } else if (args[i] === '-iF') {
        const filePath = args[i + 1];
        try {
            if (!fs.existsSync(filePath)) {
                console.error(chalk.red(`File not found: ${filePath}`));
                process.exit(1);
            }
            const fileContent = fs.readFileSync(filePath, 'utf-8');
            ipAddresses = fileContent.split(/\r?\n/).map(line => line.trim()).filter(line => line);
            
            // Check if the file contains any CIDR ranges along with IPs
            const hasCidr = ipAddresses.some(isValidIpRange);
            const hasIps = ipAddresses.some(isValidIp);

            if (hasCidr && hasIps) {
                console.log(chalk.yellow("CIDR is mixed with valid IPs, convert those to IPs."));
                process.exit(1);
            } else if (hasCidr) {
                console.log(chalk.yellow("Please provide the list of IPs, not the CIDR Notation. Can't Handle that as of now."));
                process.exit(1);
            } else if (!validateIps(ipAddresses)) {
                console.error(chalk.red("File contains invalid IP address or range format."));
                process.exit(1);
            }
            i++;
        } catch (error) {
            console.error(chalk.red(`Error reading file: ${error.message}`));
            process.exit(1);
        }
    } else if (args[i] === '-o') {
        outputFileName = args[i + 1] ? args[i + 1] : 'output.csv';
        i++;
    }
}

// Check if any IP addresses are provided
if (ipAddresses.length === 0) {
    console.error(chalk.red("No IP addresses provided. Use -i <IP-Address> or -iF <File-Path>."));
    process.exit(1);
}

// Check for CIDR notation in command-line input
const hasCidrInArgs = ipAddresses.some(isValidIpRange);
if (hasCidrInArgs) {
    console.log(chalk.yellow("I don't handle the CIDR Range."));
    process.exit(1);
}


// Section 3: DNS Lookup with Completion Tracking and Error Handling

// Function to perform reverse DNS lookup on individual IPs only
const handleDnsLookup = (ip) => {
    if (isValidIpRange(ip)) {
        // CIDR range detected; log a message indicating it can't be handled
        console.log(chalk.yellow(`Can't handle the CIDR range: "${ip}"`));
        results.push([ip, 'Not Applicable', 'CIDR range - Not handled']);
        completedLookups++;

        // Check if all lookups are complete and finalize output
        if (completedLookups === totalLookups) {
            finalizeOutput();
        }
    } else if (isValidIp(ip)) {
        // Perform DNS lookup for individual IP addresses
        dns.reverse(ip, (err, hostnames) => {
            const status = err ? 'Failed' : 'Success';
            const hostname = err ? `Error: ${err.message}` : hostnames.join(', ');

            results.push([ip, status, hostname]); // Store result for CSV export
            completedLookups++;

            // Check if all lookups are complete and finalize output
            if (completedLookups === totalLookups) {
                finalizeOutput();
            }
        });
    } else {
        // Handle invalid IP or range (shouldn’t reach here due to prior validation)
        console.error(chalk.red(`Invalid IP address or range format encountered: "${ip}"`));
        completedLookups++;
        
        // Check if all lookups are complete and finalize output
        if (completedLookups === totalLookups) {
            finalizeOutput();
        }
    }
};

// Track the total number of lookups
let totalLookups = 0;
ipAddresses.forEach(ipLine => {
    const ipParts = ipLine.split(',').map(part => part.trim());
    totalLookups += ipParts.length; // Count each IP or range separately
    ipParts.forEach(ip => handleDnsLookup(ip));
});

// Function to finalize output after all lookups are complete
function finalizeOutput() {
    printTable(); // Print results in tabular format
    if (outputFileName) saveToCSV(); // Save to CSV if specified
}


// Section 4: Table Printing and CSV Export Functions

// Function to Print Table with Proper Formatting and Color-Coded Status
function printTable() {
    if (results.length <= 1) {
        console.log(chalk.yellow("No results to display."));
        return;
    }

    // Determine the maximum column widths
    const maxIpLength = Math.max(...results.map(row => row[0].length), 'IP Address'.length);
    const maxStatusLength = Math.max(...results.map(row => row[1].length), 'Status'.length);
    const maxHostnameLength = Math.max(...results.map(row => row[2].length), 'Hostname'.length);

    // Construct row separator based on column widths
    const rowSeparator = `+${'-'.repeat(maxIpLength + 2)}+${'-'.repeat(maxStatusLength + 2)}+${'-'.repeat(maxHostnameLength + 2)}+`;

    // Print table header
    console.log(rowSeparator);
    console.log(`| ${'IP Address'.padEnd(maxIpLength)} | ${'Status'.padEnd(maxStatusLength)} | ${'Hostname'.padEnd(maxHostnameLength)} |`);
    console.log(rowSeparator);

    // Print each row with color-coded status
    results.slice(1).forEach(row => {
        // Apply color to the status and pad to ensure alignment
        const colorStatus = row[1] === 'Success' 
            ? chalk.green(row[1].padEnd(maxStatusLength)) 
            : chalk.red(row[1].padEnd(maxStatusLength));

        console.log(`| ${row[0].padEnd(maxIpLength)} | ${colorStatus} | ${row[2].padEnd(maxHostnameLength)} |`);
        console.log(rowSeparator); // Add a separator after each row
    });
}

// Function to Save Results to CSV if -o Option is Provided
function saveToCSV() {
    const csvData = results.map(row => row.join(',')).join('\n');
    const filePath = path.resolve(__dirname, outputFileName);
    fs.writeFileSync(filePath, csvData);
    console.log(chalk.green(`Results saved to ${filePath}`));
}

File: Python-Script/hosthunter.py
import argparse
import csv
import socket
import sys
import os
from colorama import Fore, Style, init

# Initialize colorama for colored output in Windows
init(autoreset=True)

def is_valid_ip(ip):
    """Validate if an IP address is valid."""
    try:
        socket.inet_aton(ip)
        return True
    except socket.error:
        return False

def is_valid_ip_range(ip):
    """Check if the input is in CIDR notation, which is unsupported."""
    return '/' in ip

def perform_dns_lookup(ip):
    """Perform reverse DNS lookup on a single IP address."""
    try:
        hostname = socket.gethostbyaddr(ip)[0]
        return "Success", hostname
    except socket.herror:
        return "Failed", "No hostname found"

def load_ips_from_file(file_path):
    """Load IP addresses from a file."""
    with open(file_path, 'r') as file:
        return [line.strip() for line in file if line.strip()]

def print_table(results):
    """Print results in a table format with color-coded status."""
    # Calculate max column widths
    max_ip_len = max(len(row[0]) for row in results)
    max_status_len = max(len(row[1]) for row in results)
    max_hostname_len = max(len(row[2]) for row in results)

    # Construct row separator based on column widths
    row_separator = f"+{'-' * (max_ip_len + 2)}+{'-' * (max_status_len + 2)}+{'-' * (max_hostname_len + 2)}+"
    print(row_separator)
    
    # Header row with the same width calculations
    header = f"| {'IP Address'.ljust(max_ip_len)} | {'Status'.ljust(max_status_len)} | {'Hostname'.ljust(max_hostname_len)} |"
    print(header)
    print(row_separator)

    for row in results:
        ip, status, hostname = row
        # Apply color to the status
        color_status = Fore.GREEN + status + Style.RESET_ALL if status == "Success" else Fore.RED + status + Style.RESET_ALL

        # Print each row with color and alignment
        print(f"| {ip.ljust(max_ip_len)} | {color_status.ljust(max_status_len + len(Fore.GREEN) + len(Style.RESET_ALL))} | {hostname.ljust(max_hostname_len)} |")
        print(row_separator)  # Add separator after each row


def save_to_csv(results, output_file):
    """Save results to a CSV file."""
    with open(output_file, 'w', newline='') as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow(['IP Address', 'Status', 'Hostname'])
        writer.writerows(results)
    print(Fore.GREEN + f"Results saved to {output_file}")

def main():
    parser = argparse.ArgumentParser(description="HostHunter - A Reverse DNS Lookup Tool")
    parser.add_argument('-i', type=str, help="Comma-separated IP addresses")
    parser.add_argument('-iF', type=str, help="Path to a file with IP addresses")
    parser.add_argument('-o', type=str, help="Output CSV file (optional)")
    args = parser.parse_args()

    ip_addresses = []
    if args.i:
        ip_addresses = [ip.strip() for ip in args.i.split(',')]
    elif args.iF:
        if not os.path.exists(args.iF):
            print(Fore.RED + f"File not found: {args.iF}")
            sys.exit(1)
        ip_addresses = load_ips_from_file(args.iF)
    else:
        print(Fore.RED + "No IP addresses provided. Use -i <IP-Address> or -iF <File-Path>.")
        sys.exit(1)

    # Check for CIDR and mixed IP/CIDR content
    has_cidr = any(is_valid_ip_range(ip) for ip in ip_addresses)
    has_ips = any(is_valid_ip(ip) for ip in ip_addresses)

    if has_cidr and has_ips:
        print(Fore.YELLOW + "Only IPs are accepted, convert the CIDR into the list of IPs and update the file.")
        sys.exit(1)
    elif has_cidr:
        print(Fore.YELLOW + "I don't handle the CIDR Range.")
        sys.exit(1)

    # Perform DNS lookup for each IP and collect results
    results = [['IP Address', 'Status', 'Hostname']]
    for ip in ip_addresses:
        if is_valid_ip(ip):
            status, hostname = perform_dns_lookup(ip)
            results.append([ip, status, hostname])
        else:
            print(Fore.RED + f"Invalid IP address format: {ip}")
            results.append([ip, "Invalid", "N/A"])

    # Print table of results
    print_table(results[1:])

    # Save to CSV if -o argument is provided
    if args.o:
        save_to_csv(results[1:], args.o)

if __name__ == "__main__":
    main()
